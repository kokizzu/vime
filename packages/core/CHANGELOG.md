# Change Log

All notable changes to this project will be documented in this file.

## [1.0.2](https://github.com/vime-js/vime/compare/@vime/core@1.0.1...@vime/core@1.0.2) (2020-08-13)


### Bug Fixes

* **core:** point unpkg and jsdelivr to correct file ([8360ad6](https://github.com/vime-js/vime/commit/8360ad62197ab35a3f51986f83e9412c7fd11a3b))

## [1.0.1](https://github.com/vime-js/vime/compare/@vime/core@1.0.0...@vime/core@1.0.1) (2020-08-13)


### Bug Fixes

* **core:** icons not loading ([2b48c85](https://github.com/vime-js/vime/commit/2b48c8547b0b7454b48a2c2707e82d74837ae9cb))

# 1.0.0 (2020-08-12)


### Bug Fixes

* **core:** lazy loading broken ([792520a](https://github.com/vime-js/vime/commit/792520ac17dcd8d70179ac4b29d91d4ac5a127eb))
* **core:** prevent posters loading twice with custom UI ([5aefb06](https://github.com/vime-js/vime/commit/5aefb06f3ad7e80204ce4966f42d5372a3517a70))
* **core/ui:** slider keydown events propagating ([94fafe1](https://github.com/vime-js/vime/commit/94fafe1f485e841febea3dd8a065588bf874622a))
* **core/ui/captions:** emitting events after disconnecting ([5f0abac](https://github.com/vime-js/vime/commit/5f0abac07f6b497607968265435e76a248ba3edf))
* **core/ui/captions:** toggling captions on/off broken ([090d776](https://github.com/vime-js/vime/commit/090d7762401750d91f9a1cc12a320cbfa3652eb6))


### Features

* **core:** add lazy loading via lozad ([3bde06c](https://github.com/vime-js/vime/commit/3bde06c5928895195f8e934c6808eca964737d07))
* **core:** add skeleton loading animation ([ff7860d](https://github.com/vime-js/vime/commit/ff7860ddfe74f029c3d365870a31d6d3d72f2748))
* **core:** dailymotion + hls + dash providers ([c520c30](https://github.com/vime-js/vime/commit/c520c302e5db91f20c9944915a43102ff1305e51))
* **core:** improve test harness + vimeo provider ([dd153ca](https://github.com/vime-js/vime/commit/dd153ca96026a3ef80c89fa1dcc560e36dcc7c1d))
* **core:** vime-embed component ([b45be6f](https://github.com/vime-js/vime/commit/b45be6f642e286b7da568b4450e8268334961019))
* **core:** vime-player component ([c86bee4](https://github.com/vime-js/vime/commit/c86bee4f709fd51d6457805df6c6c19efa0abc89))
* **core:** youtube provider + providers test harness ([8f93a68](https://github.com/vime-js/vime/commit/8f93a68edd16d4396c708e84e8e5bc1d8da69c41))
* **core/ui:** add click-to-play component ([fcce4f1](https://github.com/vime-js/vime/commit/fcce4f11b32f4603f96677dd10803233c344e8d8))
* **core/ui:** add light audio player theme ([93d4206](https://github.com/vime-js/vime/commit/93d42062ee2515ac37a10244961767714cfa87f9))
* **core/ui:** captions component ([c810f42](https://github.com/vime-js/vime/commit/c810f42d00cc81aa3cf35bf81baf2846dcbba4bb))
* **core/ui:** controls components ([5dc4e26](https://github.com/vime-js/vime/commit/5dc4e261e23df6993694abfa776af8637fb7b9a6))
* **core/ui:** create settings components ([3a1fd82](https://github.com/vime-js/vime/commit/3a1fd82a232b67f27759dc94e1e76fea8228f97e))
* **core/ui:** icons + icon component ([d13df8e](https://github.com/vime-js/vime/commit/d13df8ef1b265ec995b8a37f023162f7a978da37))
* **core/ui:** poster + spinner components ([cbdd69f](https://github.com/vime-js/vime/commit/cbdd69ff9f3474639f2a781809859c3b94beb7ff))
* **core/ui:** scrim component ([df47efe](https://github.com/vime-js/vime/commit/df47efe31b3088157896120b9f7ff2d78bdd5c99))
* **core/ui/controls:** add option to scale control size up/down ([3635b1a](https://github.com/vime-js/vime/commit/3635b1ae0d2d2d0163086a01d4e1f6fb2468a561))
* **core/ui/poster:** add poster events ([9d36f47](https://github.com/vime-js/vime/commit/9d36f47a55f6459e8ffc24018cf2fb5c2654fcde))
* **core/ui/scrim:** add events ([1180240](https://github.com/vime-js/vime/commit/1180240c8e93cce96009c3c2098064806dfa4c37))
* **core/ui/scrim:** add gradient option ([adf6bcb](https://github.com/vime-js/vime/commit/adf6bcbdb2d8520552bb6c93b5bc1f30ab95ce84))
* **core/ui/settings:** add audio/mobile styles to settings ([c937312](https://github.com/vime-js/vime/commit/c937312b344e2a5b3bb10315367f98c6b0eb81a0))
* **core/ui/spinner:** add events ([13dfd14](https://github.com/vime-js/vime/commit/13dfd1445d72357a1ed315cd3a25e955be19a772))
