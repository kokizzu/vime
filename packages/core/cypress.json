{
  "baseUrl": "http://localhost:3333/src/components",
  "watchForFileChanges": false,
  "integrationFolder": "cypress/tests",
  "testFiles": "**/*.e2e.ts",
  "fixturesFolder": false,
  "supportsFolder": "cypress/support/index.ts",
  "ignoreTestFiles": [
    "**/__snapshots__/*",
    "**/__image_snapshots__/*"
  ],
  "env": {
    "cypress-plugin-snapshots": {
      "imageConfig": {
        "threshold": 0.05
      }
    }
  }
}
