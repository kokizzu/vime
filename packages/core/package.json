{
  "name": "@vime/core",
  "version": "1.11.0",
  "description": "Collection of web components to build and customize your own media player.",
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/index.js",
  "es2015": "dist/esm/index.js",
  "es2017": "dist/esm/index.js",
  "jsnext:main": "dist/esm/index.js",
  "collection": "dist/collection/collection-manifest.json",
  "collection:main": "dist/collection/index.js",
  "types": "dist/types/index.d.ts",
  "files": [
    "dist/",
    "loader/",
    "icons/",
    "themes/"
  ],
  "contributors": [
    "Rahim Alwer <rahim_alwer@outlook.com>"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/vime-js/vime.git"
  },
  "bugs": {
    "url": "https://github.com/vime-js/vime/issues"
  },
  "scripts": {
    "build": "npm run clean:docs && stencil build --docs && npm run build:post && npm run build:themes && npm run lint:md",
    "build:post": "node scripts/post-build.js",
    "build:themes": "rimraf themes/* && npm run min:theme.default && npm run min:theme.light",
    "build:coverage": "npm run build && npm run coverage:instrument",
    "clean:docs": "rimraf ../../docs/docs/components",
    "coverage:instrument": "nyc instrument dist --in-place",
    "coverage:combine": "nyc report --reporter lcov --reporter text --report-dir coverage",
    "generate": "stencil generate",
    "generate:provider": "node scripts/new-provider.js",
    "lint:md": "npm run lint:md --prefix ../../",
    "min:theme.default": "cleancss -o themes/default.css src/globals/themes/default.css",
    "min:theme.light": "cleancss -o themes/light.css src/globals/themes/light.css",
    "serve": "concurrently \"npm run serve:public\" \"npm run serve:dev\"",
    "serve:dev": "stencil build --dev --watch --serve --port 3336",
    "serve:public": "node scripts/link-public.js && sirv public --single tests/index.html --cors --port 3334 --quiet",
    "serve:prod": "npm run build && npm run serve:public",
    "test:unit": "stencil test --spec --max-workers=2",
    "test:unit.debug": "pnpx --node-arg=\"--inspect-brk\" stencil test --spec",
    "test:unit.watch": "stencil test --spec --watchAll --max-workers=2",
    "test:e2e": "npm run build:coverage && start-server-and-test serve:public http://localhost:3334 cy:run",
    "test:e2e.ui": "npm run build && start-server-and-test serve:public http://localhost:3334 cy:run.ui",
    "test:e2e.providers": "npm run build && start-server-and-test serve:public http://localhost:3334 cy:run.providers",
    "test:ci": "rimraf .nyc_output && npm run test:unit --coverage && npm run test:e2e && npm run coverage:combine",
    "release": "semantic-release -e ../../scripts/semantic-release",
    "cy:open": "cypress open",
    "cy:run": "cypress run",
    "cy:run.ui": "cypress run --spec \"cypress/tests/ui/**/*\"",
    "cy:run.providers": "cypress run --spec \"cypress/tests/providers/**/*\""
  },
  "dependencies": {
    "stencil-wormhole": "^3.2.1"
  },
  "devDependencies": {
    "@babel/core": "^7.11.4",
    "@babel/preset-env": "^7.11.0",
    "@cypress/code-coverage": "^3.8.1",
    "@cypress/skip-test": "^2.5.0",
    "@cypress/webpack-preprocessor": "^5.4.5",
    "@stencil/angular-output-target": "^0.0.5",
    "@stencil/core": "2.0.0-8",
    "@stencil/react-output-target": "^0.0.9",
    "@stencil/sass": "^1.3.2",
    "@stencil/svelte-output-target": "^0.0.3",
    "@stencil/vue-output-target": "^0.0.3",
    "@types/hls.js": "^0.13.0",
    "@types/jest": "^26.0.10",
    "@types/puppeteer": "^3.0.1",
    "@vime/media": "^1.0.0",
    "babel-loader": "^8.1.0",
    "clean-css-cli": "^4.3.0",
    "concurrently": "^5.3.0",
    "cypress": "^4.12.1",
    "cypress-plugin-snapshots": "^1.4.4",
    "cypress-plugin-tab": "^1.0.5",
    "jest": "^26.4.2",
    "jest-watch-typeahead": "^0.6.0",
    "js-base64": "2.5.2",
    "nyc": "^15.1.0",
    "rimraf": "^3.0.2",
    "sirv-cli": "^1.0.6",
    "start-server-and-test": "^1.11.3",
    "svelte": "^3.24.1",
    "ts-loader": "^8.0.3",
    "webpack": "^4.44.1"
  },
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "vime",
    "player",
    "media",
    "audio",
    "video",
    "html5",
    "stenciljs",
    "youtube",
    "dailymotion",
    "vimeo",
    "hls",
    "dash",
    "custom",
    "web components"
  ],
  "nyc": {
    "report-dir": "cypress-coverage"
  }
}
