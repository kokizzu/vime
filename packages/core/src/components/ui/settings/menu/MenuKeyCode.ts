export enum MenuKeyCode {
  Tab = 9,
  Enter = 13,
  Esc = 27,
  Space = 32,
  PageUp = 33,
  PageDown = 34,
  End = 35,
  Home = 36,
  Left = 37,
  Up = 38,
  Right = 39,
  Down = 40,
}
