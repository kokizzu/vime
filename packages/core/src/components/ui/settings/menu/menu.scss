vime-menu {
  /**
   * @prop --menu-color: The text color within the menu.
   * @prop --menu-bg: The background color the menu.
   * @prop --menu-font-size: The font size of text within the menu.
   * @prop --menu-font-weight: The font weight of text within the menu.
   */

  display: flex;
  flex-direction: column;
  position: relative;
  overflow: hidden;
  width: 100%;
  height: auto;
  pointer-events: auto;
  color: var(--menu-color);
  background: var(--menu-bg);
  font-size: var(--menu-font-size);
  font-weight: var(--menu-font-weight);
  text-align: left;
  z-index: $z-index-menu;

  &[aria-hidden='true'] {
    display: none;
  }

  &:focus {
    outline: 0;
  }
}