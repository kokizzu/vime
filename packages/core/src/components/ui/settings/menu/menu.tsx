/* eslint-disable no-param-reassign */

import {
  h, Host, Component, Prop, Listen, Event, EventEmitter, Element, State, Watch, Method, writeTask,
} from '@stencil/core';
import { buildNoAncestorSelector } from '../../../../utils/dom';
import { isUndefined, isNull } from '../../../../utils/unit';

/**
 * @slot - Used to pass in the body of the menu which usually contains menu items, radio groups
 * and/or submenus.
 */
@Component({
  tag: 'vime-menu',
  styleUrl: 'menu.scss',
})
export class Menu {
  private shouldFocusOnOpen = false;

  private submenus!: NodeListOf<HTMLVimeMenuElement>;

  @Element() el!: HTMLVimeMenuElement;

  @State() menuItems!: NodeListOf<HTMLVimeMenuItemElement>;

  @Watch('menuItems')
  onMenuItemsChange() {
    this.menuItemsChange.emit(this.menuItems);
  }

  @State() currFocusedMenuItem = 0;

  @Watch('currFocusedMenuItem')
  async onFocusedMenuItemChange() {
    const menuItem = await this.getFocusedMenuItem();
    this.focusedMenuItem.emit(menuItem);
  }

  /**
   * Whether the menu is open/visible.
   */
  @Prop({ mutable: true, reflect: true }) active = false;

  @Watch('active')
  onActiveChange() {
    if (this.active) {
      this.findMenuItems();
      this.findSubmenus();
    }

    this.active ? this.open.emit() : this.close.emit();
  }

  /**
   * The `id` attribute of the menu.
   */
  @Prop() identifier!: string;

  /**
   * The `id` attribute value of the control responsible for opening/closing this menu.
   */
  @Prop() controller!: string;

  /**
   * Emitted when the menu is open/active.
   */
  @Event() open!: EventEmitter<void>;

  /**
   * Emitted when the menu has closed/is not active.
   */
  @Event() close!: EventEmitter<void>;

  /**
   * Emitted when the menu items present changes.
   */
  @Event() menuItemsChange!: EventEmitter<NodeListOf<HTMLVimeMenuItemElement> | undefined>;

  /**
   * Emitted when the currently focused menu item changes.
   */
  @Event() focusedMenuItem!: EventEmitter<HTMLVimeMenuItemElement | undefined>;

  componentWillLoad() {
    this.findMenuItems();
  }

  componentDidRender() {
    if (this.active && this.shouldFocusOnOpen) {
      writeTask(() => {
        this.el.focus();
        this.shouldFocusOnOpen = false;
      });
    }
  }

  /**
   * Returns the controller responsible for opening/closing this menu.
   */
  @Method()
  async getController() {
    return document.querySelector(`#${this.controller}`) as HTMLElement;
  }

  /**
   * Returns the currently focused menu item.
   */
  @Method()
  async getFocusedMenuItem() {
    return this.menuItems[this.currFocusedMenuItem];
  }

  /**
   * This should be called directly before opening the menu to set the keyboard focus on it. This
   * is a one-time operation and needs to be called everytime prior to opening the menu.
   */
  @Method()
  async focusOnOpen() {
    this.shouldFocusOnOpen = true;
  }

  private findMenuItems() {
    this.menuItems = document.querySelectorAll(
      buildNoAncestorSelector(`#${this.identifier}`, 'vime-menu', 'vime-menu-item', 5),
    );
  }

  private async focusController() {
    const controller = await this.getController();
    controller?.focus();
  }

  private focusMenuItem(index: number) {
    let boundIndex = (index >= 0) ? index : (this.menuItems.length - 1);
    if (boundIndex >= this.menuItems.length) boundIndex = 0;
    this.currFocusedMenuItem = boundIndex;
    this.menuItems[boundIndex]?.focus();
  }

  private openSubmenu() {
    const menuItem = this.menuItems[this.currFocusedMenuItem];
    if (isUndefined(menuItem)) return;
    menuItem!.click();
    writeTask(() => {
      const submenu = document.querySelector(`#${menuItem!.menu}`) as HTMLVimeMenuElement;
      submenu?.focus();
    });
  }

  private onOpen(event: Event) {
    event.stopPropagation();
    this.findMenuItems();
    this.currFocusedMenuItem = 0;
    // Prevents forwarding click event that opened the menu to menu item.
    setTimeout(() => { this.menuItems[this.currFocusedMenuItem]?.focus(); }, 10);
    this.active = true;
  }

  private onClose() {
    this.currFocusedMenuItem = -1;
    this.active = false;
    this.focusController();
  }

  private onClick(event: Event) {
    event.stopPropagation();
  }

  private onKeyDown(event: KeyboardEvent) {
    if (!this.active || this.menuItems.length === 0) return;
    event.preventDefault();
    event.stopPropagation();
    switch (event.key) {
      case 'Escape':
        this.onClose();
        break;
      case 'ArrowDown':
      case 'Tab':
        this.focusMenuItem(this.currFocusedMenuItem + 1);
        break;
      case 'ArrowUp':
        this.focusMenuItem(this.currFocusedMenuItem - 1);
        break;
      case 'ArrowLeft':
        this.onClose();
        break;
      case 'ArrowRight':
      case 'Enter':
      case ' ':
        this.openSubmenu();
        break;
      case 'Home':
      case 'PageUp':
        this.focusMenuItem(0);
        break;
      case 'End':
      case 'PageDown':
        this.focusMenuItem(this.menuItems!.length - 1);
        break;
    }
  }

  private findSubmenus() {
    this.submenus = document.querySelectorAll(
      buildNoAncestorSelector(`#${this.identifier}`, 'vime-menu', 'vime-menu', 4),
    );
  }

  private isValidSubmenu(submenu: HTMLElement | null) {
    if (isNull(submenu)) return false;
    return !isUndefined(Array.from(this.submenus).find((menu) => menu.id === submenu!.id));
  }

  private toggleSubmenu(submenu: HTMLVimeMenuElement, isActive: boolean) {
    if (!this.isValidSubmenu(submenu)) return;

    Array.from(this.menuItems)
      .filter((menuItem) => menuItem.identifier !== submenu.controller)
      .forEach((menuItem) => { menuItem.hidden = isActive; });

    submenu.active = isActive;
  }

  @Listen('open')
  onSubmenuOpen(event: CustomEvent<void>) {
    const submenu = event.target as HTMLVimeMenuElement;
    this.toggleSubmenu(submenu, true);
  }

  @Listen('close')
  onSubmenuClose(event: CustomEvent<void>) {
    const submenu = event.target as HTMLVimeMenuElement;
    this.toggleSubmenu(submenu, false);
  }

  @Listen('click', { target: 'window' })
  onWindowClick() {
    if (this.active) this.active = false;
  }

  @Listen('keydown', { target: 'window' })
  onWindowKeyDown(event: KeyboardEvent) {
    if (this.active && (event.key === 'Escape')) this.onClose();
  }

  render() {
    return (
      <Host
        id={this.identifier}
        role="menu"
        tabindex="-1"
        aria-labelledby={this.controller}
        aria-hidden={!this.active ? 'true' : 'false'}
        onFocus={this.onOpen.bind(this)}
        onClick={this.onClick.bind(this)}
        onKeyDown={this.onKeyDown.bind(this)}
      >
        {/* Seems like without this div unnecessary re-renders keep happening. */}
        <div>
          <slot />
        </div>
      </Host>
    );
  }
}
