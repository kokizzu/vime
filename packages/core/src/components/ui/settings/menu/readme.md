# vime-menu

This component is responsible for containing and managing menu items and submenus. The menu is ARIA
friendly by ensuring the correct ARIA properties are set, and enabling keyboard navigation when it
is focused.

You can use this component if you'd like to build out a custom settings menu. If you're looking
to only customize the content of the settings see [`vime-settings`](../settings/readme.md), and if
you want an easier starting point see [`vime-default-settings`](../default-settings/readme.md).

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/settings/menu/menu.png"
  alt="Vime settings menu component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {5-11} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-menu
      identifer="menu-id"
      controller="menu-controller-id"
      [active]="isMenuActive"
    >
      <!-- ... -->
    </vime-menu>
  </vime-ui>
</vime-player>
```

```ts title="example.ts"
class Example {
  isMenuActive = false;

  // ...
}
```

### Html

```html {5-7}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-menu identifer="menu-id" controller="menu-controller-id" active>
      <!-- ... -->
    </vime-menu>
  </vime-ui>
</vime-player>
```

### React

```tsx {2,12-18}
import React, { useState } from "react";
import { VimePlayer, VimeUi, VimeMenu } from "@vime/react";

function Example() {
  const [isMenuActive, setIsMenuActive] = useState(false);

  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeMenu
          identifer="menu-id"
          controller="menu-controller-id"
          active={isMenuActive}
        >
          <!-- ... -->
        </VimeMenu>
      </VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {6-12,18,24} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeMenu
        identifer="menu-id"
        controller="menu-controller-id"
        :active="isMenuActive"
      >
        <!-- ... -->
      </VimeMenu>
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi, VimeMenu } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeMenu,
    },

    data: {
      isMenuActive: false,
    },
  };
</script>
```

## Properties

| Property                  | Attribute    | Description                                                                        | Type      | Default     |
| ------------------------- | ------------ | ---------------------------------------------------------------------------------- | --------- | ----------- |
| `active`                  | `active`     | Whether the menu is open/visible.                                                  | `boolean` | `false`     |
| `controller` _(required)_ | `controller` | The `id` attribute value of the control responsible for opening/closing this menu. | `string`  | `undefined` |
| `identifier` _(required)_ | `identifier` | The `id` attribute of the menu.                                                    | `string`  | `undefined` |

## Events

| Event                  | Description                                           | Type                                                            |
| ---------------------- | ----------------------------------------------------- | --------------------------------------------------------------- |
| `vClose`               | Emitted when the menu has closed/is not active.       | `CustomEvent<void>`                                             |
| `vFocusMenuItemChange` | Emitted when the currently focused menu item changes. | `CustomEvent<HTMLVimeMenuItemElement \| undefined>`             |
| `vMenuItemsChange`     | Emitted when the menu items present changes.          | `CustomEvent<NodeListOf<HTMLVimeMenuItemElement> \| undefined>` |
| `vOpen`                | Emitted when the menu is open/active.                 | `CustomEvent<void>`                                             |

## Methods

### `focusOnOpen() => Promise<void>`

This should be called directly before opening the menu to set the keyboard focus on it. This
is a one-time operation and needs to be called everytime prior to opening the menu.

#### Returns

Type: `Promise<void>`

### `getController() => Promise<HTMLElement>`

Returns the controller responsible for opening/closing this menu.

#### Returns

Type: `Promise<HTMLElement>`

### `getFocusedMenuItem() => Promise<HTMLVimeMenuItemElement>`

Returns the currently focused menu item.

#### Returns

Type: `Promise<HTMLVimeMenuItemElement>`

## Slots

| Slot | Description                                                                                           |
| ---- | ----------------------------------------------------------------------------------------------------- |
|      | Used to pass in the body of the menu which usually contains menu items, radio groups and/or submenus. |

## CSS Custom Properties

| Name                 | Description                              |
| -------------------- | ---------------------------------------- |
| `--menu-bg`          | The background color the menu.           |
| `--menu-color`       | The text color within the menu.          |
| `--menu-font-size`   | The font size of text within the menu.   |
| `--menu-font-weight` | The font weight of text within the menu. |

## Dependencies

### Used by

- [vime-settings](../settings)
- [vime-submenu](../submenu)

### Graph

```mermaid
graph TD;
  vime-settings --> vime-menu
  vime-submenu --> vime-menu
  style vime-menu fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
