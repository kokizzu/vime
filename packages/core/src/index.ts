import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
