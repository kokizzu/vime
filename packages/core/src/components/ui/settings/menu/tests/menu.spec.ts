import { SpecPage, newSpecPage } from '@stencil/core/testing';
import { Menu } from '../menu';

let page: SpecPage;

beforeEach(async () => {
  page = await newSpecPage({
    components: [Menu],
    html: '<vime-menu />',
  });
});

it('should be structurally sound', () => {
  expect(page.root).toMatchSnapshot();
});

// More advanced menu tests are performed as E2E tests.
