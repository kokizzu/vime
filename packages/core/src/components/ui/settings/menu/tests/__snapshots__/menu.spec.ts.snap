// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-menu aria-hidden="true" role="menu" tabindex="-1">
  <!---->
  <div></div>
</vime-menu>
`;
